tensorflow/java/src/main/java/org/tensorflow/Graph.java

```diff
@@ -143,6 +143,82 @@ public void importGraphDef(byte[] graphDef, String prefix) throws IllegalArgumen
     }
   }
 
+  /**
+   * Adds operations to compute the partial derivatives of sum of {@code y}s w.r.t {@code x}s,
+   * i.e., {@code d(y_1 + y_2 + ...)/dx_1, d(y_1 + y_2 + ...)/dx_2...}
+   * <p> 
+   * {@code dx} are used as initial gradients (which represent the symbolic partial derivatives of some loss function 
+   * {@code L} w.r.t. {@code y}). {@code dx} must be null or have size of {@code y}.
+   * <p>
+   * If {@code dx} is null, the implementation will use dx of {@link org.tensorflow.op.core.OnesLike OnesLike} for all
+   * shapes in {@code y}.
+   * 
+   * @param y output of the function to derive
+   * @param x inputs of the function for which partial derivatives are computed
+   * @param dx if not null, the partial derivatives of some loss function {@code L} w.r.t. {@code y}
+   * @return the partial derivatives {@code dy} with the size of {@code x}
+   */
+  public Output<?>[] addGradients(Output<?>[] y, Output<?>[] x, Output<?>[] dx) {
```

Comment:
I wasn't sure about this, you probably know better than me: is it possible to have a graph with variables of different types? If it is guaranteed that could never happen, I'll gladly remove those wildcards to enforce type-safety.

Reply:
Hmm...it actually is, since one could have say `tf.cast`s on the path between `x` and `y`, something like:

```python
import tensorflow as tf

x = tf.placeholder(tf.float64)
y = tf.square(tf.cast(x, tf.float32))
dy = tf.gradients(y, x)[0]
print(x.dtype, y.dtype, dy.dtype)
```

So apologies, ignore my comment :)